src/differencer/jsonDiff.ts

```diff
@@ -0,0 +1,382 @@
+    textDiff: {
+      // default 60, minimum string length (left and right sides) to use text diff algorithm: google-diff-match-patch
+      minLength: 6000
+    }
+  });
+  //add plugin to include ID of the JSON node in the diff
+  jsonDiff.processor.pipes.diff.before("collectChildren", addDiffId);
+  ramlToolLogger.debug(
+    `Added plugin to include node ID in the diff: ${jsonDiff.processor.pipes.diff.list()}`
+  );
+
+  const diffs = jsonDiff.diff(left, right);
+
+  return parseDiffs(diffs);
+}
+
+/**
+ * Verify object to diff/compare conform to flattened AMF graph json structure
+ * @param obj Json provided to diff/compare
+ * @param objLoc Left/Right json in the diff
+ */
+function validateJson(obj: any, objLoc: "left" | "right"): void {
+  let msg;
+  if (obj == null || _.isEmpty(obj)) {
+    msg = `Invalid ${objLoc} object`;
+    logError(msg);
+    return;
+  }
+  //Flattened AMF graph must have valid @graph or @context properties
+  const graph = obj[JSON_LD_KEY_GRAPH];
+  if (graph == null) {
+    msg = `${_.startCase(objLoc)} json must have ${JSON_LD_KEY_GRAPH} property`;
+    logError(msg);
+    return;
+  }
+  if (!_.isArray(graph) || graph.length == 0) {
+    msg = `${JSON_LD_KEY_GRAPH} property in ${objLoc} json must be an array of json nodes`;
+    logError(msg);
+    return;
+  }
+}
+function logError(msg: string): never {
+  if (msg == null) {
+    msg = "Unknown error in json diff";
+  }
+  ramlToolLogger.error(msg);
+  throw new Error(msg);
+}
+/**
+ * Parse JSON differences and generate an array of typed diff objects
+ * @param diffs Differences JSON
+ *
+ * @returns Array of NodeDiff objects
+ */
+function parseDiffs(diffs: object): Array<NodeDiff> {
+  const typedDiffs: Array<NodeDiff> = [];
+  if (diffs == null || _.isEmpty(diffs)) {
+    //no changes
+    ramlToolLogger.info("No differences found by the jsondiffPatch");
+    return typedDiffs;
+  }
+
+  ramlToolLogger.debug("Parsing differences generated by jsondiffPatch");
+  const graphDiffs = diffs[JSON_LD_KEY_GRAPH];
+  if (graphDiffs != null) {
+    //process graph changes
+    ramlToolLogger.debug("Parsing differences in JSON-LD graph node");
+    typedDiffs.push(...parseGraph(graphDiffs));
+  }
+  const contextDiffs = diffs[JSON_LD_KEY_CONTEXT];
+  if (contextDiffs != null) {
+    //process context changes
+    ramlToolLogger.debug("Parsing differences in the JSON-LD context node");
+    const typedContextDiff = parseNodePropDiffs(
+      JSON_LD_KEY_CONTEXT,
+      contextDiffs
+    );
+    if (typedContextDiff != null) {
+      typedDiffs.push(typedContextDiff);
+    }
+  }
+  return typedDiffs;
+}
+
+/**
+ * Parse differences for each JSON node in the graph
+ * @param graphDiffs Array of JSON nodes with differences
+ *
+ * @returns Array of NodeDiff objects
+ */
+function parseGraph(graphDiffs: Array<object>): Array<NodeDiff> {
+  const typedDiffs: Array<NodeDiff> = [];
+  Object.keys(graphDiffs).forEach(key => {
+    if (key === JSON_DIFF_PATCH_ARRAY_KEY) {
```

Comment:
Are there non-numeric keys that you want to iterate over? If there are, please add a comment describing them. If not, then you don't need to use `Object.keys` or this `if` block.

Reply:
It is little complicated to explain in a comment. I will add a brief one. But the best is to look at the jsondiffpatch doc